### PR TITLE
feat(auth): Decode Nhost JWT and add in needed claims and roles for API side

### PR DESCRIPTION
**Pull request by dthyresson**
[Nhost](https://nhost.io/) auth lacked a decoder, therefore just the token was returned.

This PR adds decoding and verifying via the secret and also injects several claims not on the decoded token that are helpful for use on RedwoodJS's API side: sub (user id), issuer (nhost), and the roles.

Without the decoded, it was returning the token itself, unverified and not decoded.

The raw Nhost decoded token is just:

```js
{
  "https://hasura.io/jwt/claims": {
    "x-hasura-user-id": "2dd1bd63-d8e0-4ad1-be59-8edfd4e5935a",
    "x-hasura-allowed-roles": [
      "user",
      "me"
    ],
    "x-hasura-default-role": "user"
  },
  "iat": 1619270071,
  "exp": 1619270971
}
```

Notice there is no sub, iss, or roles but this info is in the "hasura claims namespace".

With this decoder, now the result is a decoded token set in currentUser of:

```js
"currentUser": {
    "decoded": {
      "https://hasura.io/jwt/claims": {
        "x-hasura-user-id": "2dd1bd63-d8e0-4ad1-be59-8edfd4e5935a",
        "x-hasura-allowed-roles": [
          "user",
          "me"
        ],
        "x-hasura-default-role": "user"
      },
      "iat": 1619270071,
      "exp": 1619270971,
      "sub": "2dd1bd63-d8e0-4ad1-be59-8edfd4e5935a",
      "iss": "nhost",
      "roles": [
        "user",
        "me"
      ]
    },
    "type": "nhost",
    "token": "jwt"
  },
```

This can be used on the api side to easily fetch the user id from the traditional sub claim and also enforce api side RBAC via requireAuth() and hasRoles().

### CODE REVIEW

**Comment by dthyresson** [comment]
Note if approved will need to update Nhost auth docs to explain required and optionsal envars:

```
NHOST_JWT_SECRET="your-secret-that-you-keep"
NHOST_BACKEND_URL="https://backend-XXXXXXX.nhost.app"

# OPTIONAL
NHOST_CLAIMS_NAMESPACE="https://hasura.io/jwt/claims"
NHOST_SUB_CLAIM="x-hasura-user-id"
NHOST_ROLES_CLAIM="x-hasura-allowed-roles"
```

**Review by Tobbe — APPROVED** [review]

**Comment by Tobbe** [comment]
@dthyresson This looks good to me, I approved it. But I think we should let @nunopato have a look too, before merging

**Comment by dthyresson** [comment]
> @dthyresson This looks good to me, I approved it. But I think we should let @nunopato have a look too, before merging

Sounds good. I have been chatting with @nunopato on Nhost's Discord and referenced the issue and reason for the decoder here in their HBP repo: https://github.com/nhost/hasura-backend-plus/issues/495

File: packages/cli/src/commands/setup/auth/providers/nhost.js
Comment (dthyresson) [inline packages/cli/src/commands/setup/auth/providers/nhost.js:21]: Note I added the NHOST_JWT_SECRET requirement to setup command.

**Comment by nunopato** [comment]
@dthyresson thanks for putting this up. Are you waiting for the work on nhost/hasura-backend-plus#495 to be released or will you merge this one already?

**Review by nunopato — APPROVED** [review]

**Comment by dthyresson** [comment]
> @dthyresson thanks for putting this up. Are you waiting for the work on [nhost/hasura-backend-plus#495](https://github.com/nhost/hasura-backend-plus/issues/495) to be released or will you merge this one already?

@nunopato I think these changes would still work even with a fix for https://github.com/nhost/hasura-backend-plus/issues/49 -- likely they would simply rewrite the sub, iss, roles claims. But I'll defer to you if you think it best to wait.

In any case, we will 100% need the decoder w/ the secret -- so perhaps we merge and then update in 0.32 if needed after the fix. Let's do that so can at least get started on using Nhost auth with Redwood in 0.31.

**Comment by nunopato** [comment]
@dthyresson Sounds good! as you point out, we can update it afterward in 0.32.

**Comment by dthyresson** [comment]
FYI - Nhost added sub and issuer 

https://github.com/nhost/hasura-backend-plus/pull/496

but will still be ok to override I think. Or, @nunopato should I just keep the roles in?

I guess it depends when https://github.com/nhost/hasura-backend-plus/pull/496 goes live. If it is w/in a day or so, then I'll rework this PR to remove those claims.

**Comment by nunopato** [comment]
@dthyresson I think it is best if you don't wait. We still have some work to do in preparation for the release which will include nhost/hasura-backend-plus#496.

**Comment by nunopato** [comment]
@dthyresson anything pending from your side to have this one released? let me know if I can help in any way.

**Comment by dthyresson** [comment]
> @dthyresson anything pending from your side to have this one released? let me know if I can help in any way.

Thanks @nunopato I'll confirm the new claims and work to get this in for the 0.32 release next week,

**Comment by dthyresson** [comment]
> let me know if I can help in any way.

@nunopato I just logged in using our https://redwood-playground-auth.netlify.app/nhost

```
{
  "https://hasura.io/jwt/claims": {
    "x-hasura-user-id": "00xxxb5a-xxxx-45a0-xxxx-18a7264xxxx6",
    "x-hasura-allowed-roles": [
      "me",
      "user"
    ],
    "x-hasura-default-role": "user"
  },
  "iat": 1620215237,
  "exp": 1620216137
}

But I don't see the claims. Would I need the latest nhost sdk?

We have `nhost-js-sdk@^3.1.0` currently

**Comment by nunopato** [comment]
@dthyresson could you let me know the name of the project and the email address of the account?

**Comment by nunopato** [comment]
@dthyresson Last Friday I upgraded a project you have in your account but it was not the right one. The correct one is now running on HBP v2.4.0.

```{
  "https://hasura.io/jwt/claims": {
    "x-hasura-user-id": "d8daf6a3-2b33-4248-be81-002e4db629b7",
    "x-hasura-allowed-roles": [
      "user",
      "me"
    ],
    "x-hasura-default-role": "user"
  },
  "sub": "d8daf6a3-2b33-4248-be81-002e4db629b7",
  "iss": "nhost",
  "iat": 1620296183,
  "exp": 1620297083
}```

**Comment by dthyresson** [comment]
> The correct one is now running on HBP v2.4.0.

Perfect! Yes I see that the access token has those claims when I use https://redwood-playground-auth.netlify.app/nhost.

I'll get this PR ready for the next release.

Thanks much!

**Comment by dthyresson** [comment]
@Tobbe I think we are all good now to merge -- as I can see the additional iss/sub claims in the token -- and then test the canary with playground auth.

**Comment by Tobbe** [comment]
@dthyresson Yeah, LGTM!

**Pull request merged**
